crates/driver/src/boundary/settlement.rs

```diff
@@ -81,28 +82,45 @@ impl Settlement {
         );
 
         for trade in &solution.trades {
-            let (boundary_order, executed_amount) = match trade {
+            let (boundary_order, execution) = match trade {
                 competition::solution::Trade::Fulfillment(trade) => {
                     // TODO: The `http_solver` module filters out orders with 0
                     // executed amounts which seems weird to me... why is a
                     // solver specifying trades with 0 executed amounts?
-                    if eth::U256::from(trade.executed).is_zero() {
+                    if eth::U256::from(trade.execution.filled).is_zero() {
                         return Err(Error::Boundary(anyhow!("unexpected empty execution")));
                     }
 
-                    (to_boundary_order(&trade.order), trade.executed.into())
+                    let solver_fee = match trade.order.solver_determines_fee() {
+                        true => trade.execution.fee.expect("API ensures fee exists").0,
```

Comment:
Where is that ensured?

Reply:
That happens [here](https://github.com/cowprotocol/services/pull/1366/files#diff-9f2c31c10679a7e214aef5191d3b280418861f993fc0fa3d7da78ac8f9d994f5R45-R48). If a solver provides a trade execution without a fee for a partially fillable limit order the JSON response fails to convert to the domain `Solution` type.

Reply:
Would be nice if the invariant is kept local to the `competition::solution::Trade` type and module (i.e. not leak here).

I think I remember you mentioning that this was tricky to do, just food for thought.

Reply:
Mhh, I guess something like this could work:
```rust
pub struct SolverFee(pub competition::order::SellAmount);
pub struct ProtocolFee(pub competition::order::SellAmount);

pub enum Fee {
    Solver(SolverFee),
    Protocol(ProtocolFee),
}

impl OrderExecution {
    /// Panics when the trade should not or does not contain a [`SolverFee`].
    pub fn solver_fee(&self, order: &competition::Order) -> SolverFee {
        match (order.solver_determines_fee(), self.fee) {
            (true, Fee::Solver(fee)) => fee,
            _ => panic!("invariants violated"),
        }
    }

    /// Panics when the trade should not or does not contain a [`ProtocolFee`].
    pub fn protocol_fee(&self, order: &competition::Order) -> ProtocolFee {
        match (order.solver_determines_fee(), self.fee) {
            (true, Fee::Protocol(fee)) => fee,
            _ => panic!("invariants violated"),
        }
    }
}
```

Reply:
I did something similar by making the `Fulfillment` fields private and verifying the invariants at creation time.